### PR TITLE
Complete backend rework of the EIG

### PR DESCRIPTION
**Pull request by C0bra5**
# The goals of this rework:
1. Accurately simulate the drop rates of IC2 crops.
2. Increase modularity to increase the ease of integration for future content.
3. Batch identical seeds together and significantly reduce the NBT footprint of the EIG.
4. General QoL improvements.
5. Potentially breaking changes and the steps taken to prevent them.
6. Not affect balance, ~~for now~~

# Accurately simulate the drop rates of IC2 crops.

It took me a couple of days, but I've finally created an extremely accurate method to pre-calculate the drop rates and growth times of IC2 crops. I never thought I would need that old college statistics textbook again, but it finally became helpful after all those years!

## The problems with approximating growth speed

The most straightforward assumption would be that we have to get the average value returned by the `calcGrowthRate` and divide the growth time of the crop by that average. Unfortunately, this assumption will result in some wildly inaccurate results. Here's why:
- IC2 Crops grow in stages, each with a different growth time.
  - Until recently, this wasn't considered, often causing the EIG to use wildly inaccurate growth times.
- When a crop finishes a stage, it does not carry the excess growth to the new stage; it restarts at 0.- Crops do not grow every tick; they grow once every 256 ticks (or whatever the value of `TileEntityCrop.tickRate` is set to)
- The stage at which a crop returns after being harvested isn't always the first.
  - Sometimes, it's random, and sometimes, it's a fixed value above 1.
  - Until recently, this wasn't considered, often causing the EIG to use wildly inaccurate growth times.
- Calculating the average number of cycles as the average of all potential growth speeds of a crop is incredibly inaccurate.
  - The solution to this issue seemed easy to figure out at first, but it took the most time to figure out.
  - The problem boils down to this question:
    - What is the average number of dice throws needed to reach a given total?
  - The dice we are rolling aren't just your regular 6-sided dice either:
    - It's a fair dice (all sides are equally likely to roll), but the number on a face might be repeated on another side.  
    - The numbers do not follow each other or have a predictable linear rise.
    - The dice usually has seven faces, but if IC2 were to be modified, it could effectively have any number of faces.

### How did I approximate growth speed?

First, I generate a frequency table to help me determine how often a crop will return to a specific growth stage after being harvested.
- First, I generate a list of all possible growth speeds using a custom version of the calcGrowthRate function found in `TileEntityCrop`, which lets me control the random value.
- Then, I get the growth times for all relevant growth stages.
- Then, I calculate the average number of growth cycles needed for the crop to reach maturity.
  - See https://github.com/C0bra5/KubaTech/blob/3474776f5f1056c037a2c156ed9dd49917cbcc18/src/main/java/kubatech/tileentity/gregtech/multiblock/eigbuckets/EIGIC2Bucket.java#L556-L587
- Then, I multiply the number of rolls each growth stage takes by how often they occur.
- Finally, I calculate the average number of growth cycles needed to grow to maturity.

The above explanation could be more accurate, but it should be good enough to give you a gist of what's going on. If you want a complete understanding of how this is calculated, consider reading the code: https://github.com/C0bra5/KubaTech/blob/3474776f5f1056c037a2c156ed9dd49917cbcc18/src/main/java/kubatech/tileentity/gregtech/multiblock/eigbuckets/EIGIC2Bucket.java#L498-L542

I'm confident that my algorithm generates growth speed approximations that are ~99.95% of the actual growth speed of IC2 crops planted in a crop stick. 

## The problems with approximating drop rates of a given item

Approximating the drop rates was more challenging than anticipated, but that good old statistics textbook again proved invaluable. 

To compute the average number of items dropped per harvest, we must compute two values:
- The average number of separate item drops created per harvest.
- The average stack size increase caused by the gain stat.

### Approximating the average number of separate item drops.

When the `harvest_automated` method is called, this is how IC2 determines the number of items to drop:
```java
double chance = CropCard.dropGainChance(this) * Math.pow(1.03, getGain());
int dropCount = (int) Math.max(0, Math.round(Random.nextGaussian() * chance * 0.6827 + chance))
```

I can approximate the average number of item drops created by harvests by taking samples from a normal distribution at all points within a reasonable range using small enough intervals. I then use those samples to create a weighted average for executing the two lines of code shown above.

Essentially, I'm just evaluating this formula
```math
\overline{x}=\frac{\int_{-\infty}^{\infty}\max\left(0,\mathrm{round}\left(x\cdot a\cdot0.6827+a\right)\right)\cdot\frac{\exp\left(-{1 \over 2}x^2\right)}{\sqrt{2\pi}}, dx}{\int_{-\infty}^{\infty}\frac{\exp\left(-{1 \over 2}x^2\right)}{\sqrt{2\pi}}, dx}
```
Where
- $\overline{x}$ is the average number of drops per harvest.
- $a$ is the result of the first line of code.
- $x$ is a possible value of the nextGausian function.

Notes regarding the implementation:
- The major division is skipped since the integral of the standard normal distribution is 1.
- I'm using the standard normal distribution equation $\frac{\exp\left(-{1 \over 2}x^2\right)}{\sqrt{2\pi}}$ instead of the Gaussian equation $\exp(-x^2)$. The Gaussian formula gives inaccurate results.
- I'm using the composite trapezoidal rule to perform the integration:
  $$\int _{a}^{b}f(x)\,dx\approx {\frac {b-a}{n}}\left({\frac{f(a) + f(b)}{2}}+\sum _{k=1}^{n-1}\left(f\left(a+k{\frac {b-a}{n}}\right)\right)\right)$$

Where
- $a$ is the lower bound of the range: -10
- $b$ is the upper bound of the range: 10
- $n$ is the number of steps: 10,000
- $f(x)$ is the function to integrate.

The result of this approximation tends to be ~99.995% accurate to the average number of drops after 10 million simulations.

### Approximating stack size increased caused by gain.

When IC2 generates a drop, it rolls a number between 0 and 99 (inclusive) and adds 1 to the stack size if the random number is less than or equal to the seed's gain stat.

For once, the solution to this problem is simple: we can do `(gain+1) / 100` to compute the average increase.

### The problem with finding the individual item drop frequency of a crop.

The item drop frequency approximations have to be based on the results of calling the crop card's getGain function since IC2 crops don't natively have a way to tell us their drop rates for each item they can drop. I plan on addressing this issue in the long term by adding exactly that to most of the IC2 crops found in GTNH. Native IC2 crops must be handled separately since I can't submit changes to IC2.

## Accurately evaluating the environmental factors for the crops inside the EIG.

I've also added some functions that help me more effectively compute the EIG's environment. It uses the humidity and nutrient bonus of the biome located where the EIG controller is instead of an arbitrary point inside the EIG. It also uses the controller's y value to compute the air quality height factor. Each method uses a set of specific values that I hope to move to a configuration file that can be found here: https://github.com/C0bra5/KubaTech/blob/3474776f5f1056c037a2c156ed9dd49917cbcc18/src/main/java/kubatech/tileentity/gregtech/multiblock/eigbuckets/EIGIC2Bucket.java#L49

## How accurate are my approximations?

According to my testing, the rates I generate are effectively indistinguishable from the rates at which actual IC2 crops grow.

Here is the result of a 16-hour test. I used a custom version of GT++, where I removed the bonus from the crop managers to help me get accurate results. I tried using IC2 crop harvesters, but they don't harvest fast enough to be a precise simulation, usually lagging 2-3% behind.
image

I've updated the relevant unit tests to verify the accuracy of the EIG more effectively. Additional tests could be added since most methods used to calculate my approximations can be completely decoupled.

# Increase modularity to increase the ease of integration for future content.

The EIG used to rely on a class called `GreenhouseSlot` to store seeds. That class needed a good dose of abstraction since it used entirely different logic depending on whether the EIG was in regular or IC2 mode.

The new `EIGBucket` system adds that well-needed abstraction. Each bucket can detect whether it is compatible with a given seed and regenerates its drops in a manner that is most effective for a given seed type. If a bucket is modified, we can have the EIG reroll the drop tables by changing the revision number of that bucket.

Each bucket has its own factory class, used to create a new bucket of this type and restore a bucket from NBT.

Instances of bucket factories are stored inside an `EIGMode` class, which acts as a container for those factories. The EIG stores an instance of the relevant mode class and uses it to create new buckets and pool relevant mode-specific balance factors. Each mode also generates its tooltip section based on its balance factors. (no more hard-coded tooltip values).

The factories are loaded inside the EIG mode during the post-init event, and other mods should be able to add their bucket factories to a specific mode or even add their own EIG mode down the line (This is mostly planning for Crops-NH).

# Batch identical seeds together and reduce the NBT footprint of the EIG

Currently, if you have 3000 of the same IC2 seed or 200 stacks of the same regular seed, the EIG stores each item or regular seed stack separately. This means that high-tier EIGs with hundreds of seeds are massive NBT bombs. By batching identical seeds together, we reduce the amount of NBT stored inside the EIG controller and make it easier to generate drops, reducing the overall TPS usage.

## Changes to how seed capacity is handled.

Before, each slot had a specific number of seeds that could be stored inside (1 for IC2 and 64 for normal mode). This meant that unless you were using full stacks while using the regular mode, you were losing potential productivity. I suspect that the implementation of this limitation mainly resulted from how the slot mechanic was implemented.

The new `EIGBucket` system has an effectively infinite seed capacity per bucket, so I've taken the liberty of reworking the way storage limitations are handled in the eig:
- The slot limit controls the number of separate seed types that the EIG can store.
- The seed capacity controls the total amount of seeds the EIG can store.

This change should make the eig a little bit more flexible:

Normal Mode | IV power | 128 seed cap | max of 2 seed types
image
IC2 Mode | UIV power | Seed capacity: 4096 | Seed type limit: 4096
image

# General QoL changes
- As mentioned before, the drop rates of IC2 crops have been effectively fixed.
- It is now impossible to void seeds when the EIG is set to output mode or when the controller is broken.
- The eig tooltips now distinguishes seed type overflow from seed count overflow.
- The tooltip for the EIG has been updated to mention the stats of each mode EIG more effectively.
- Added a mention of the 5x Weed-EX 9000 usage multiplier for the IC2 mode.
- You no longer get that weird desync while using the UI to insert crops inside the EIG.
- Partial stacks of seeds appear as a single stack in the UI.
- Extracting a seed that needs a block under it to grow now also returns the block when using the UI to extract the seed.
- Support for inverted purple has been fixed.
- Rainbow Cactus can now be grown inside the EIG (consider it an example of how to implement seed-specific buckets).

# Potentially breaking changes and the steps taken to prevent them

Since I'm using a different way to store and simulate crops, I must migrate the old `GreenhouseSlot` system to the new `EIGBucket` system.
- If this update goes live, all existing EIGs will have their drop tables updated.
- I fixed the growth calculations of some crops from crops++ a few months back. Given how the EIG now handles its environmental stats, some crops will likely be impossible to grow. A good example would be the primordial pearl crop, which had its environmental stat requirements essentially quadrupled.
- Crops that cannot grow inside the EIG will be outputted automatically.
- If a seed cannot be grown inside an EIG after an update, it will also be ejected.
- To prevent unforeseen consequences, I've ensured that ejected seeds are never voided.

### CODE REVIEW

File: src/main/java/kubatech/api/enums/GTVoltageIndex.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package kubatech.api.enums;
```
Comment (OneEyeMaker) [inline src/main/java/kubatech/api/enums/GTVoltageIndex.java]: Such constants should probably live [here](https://github.com/GTNewHorizons/GT5-Unofficial/blob/master/src/main/java/gregtech/api/enums/TierEU.java).

File: src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java [hunk 2]
```diff
@@ -158,36 +149,65 @@ public class GT_MetaTileEntity_ExtremeIndustrialGreenhouse
...
+    public static final int EIG_BALANCE_WEED_EX_USAGE_BEGINS_AT = 1000;
+    public static final int EIG_BALANCE_WATER_USAGE_PER_SEED = 1000;
 
-    /***
-     * All crops above this threshold will die if WEED-EX 9000 isn't supplied.
-     * Consider changing this value after making changes to the EIG_BALANCE_IC2_ACCELERATOR_TIER.
-     */
-    private static final int EIG_CROP_LIMIT_FOR_WEEDEX9000_REQUIREMENT = 1000;
+    private static final Fluid WEEDEX_FLUID = Materials.WeedEX9000.mFluid;
+    private static final Item FORESTRY_FERTILIZER_ITEM = GameRegistry.findItem("Forestry", "fertilizerCompound");
```
Comment (OneEyeMaker) [inline src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java]: Attempting to search mod item on class load may fail, cause there's no guarantees, that mod is loaded at that point.

File: src/main/java/kubatech/api/enums/GTVoltageIndex.java (hunk 1 above)
Comment (C0bra5) [reply src/main/java/kubatech/api/enums/GTVoltageIndex.java]: Yeah I though so too. It's odd it wasn't there already. I'll make a PR to add it to that package in GT5u. It if it gets accepted I'll remove the class update the relevant classes.

Edit PR has been made: https://github.com/GTNewHorizons/GT5-Unofficial/pull/2578

File: src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java (hunk 2 above)
Comment (C0bra5) [reply src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java]: This is how it used to work; I just grouped it with the rest of the balance-related constants.
https://github.com/GTNewHorizons/KubaTech/blob/a17f0c238deadc7b107b37a5a852387a267cbb1f/src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java#L189

I've moved fertilizer item declarations to the post-init recipe loader, so it should be fine now.

**Comment by eigenraven** [comment]
Moved over to https://github.com/GTNewHorizons/GT5-Unofficial/pull/2616

**Pull request closed without merging**